### PR TITLE
feat: 457 refactor directanalysisrunner

### PR DESCRIPTION
**Pull request by Carsopre**
## Issue addressed
Solves #457 

## Code of conduct
- [x] I HAVE NOT added sensitive or compromised (test) data to the repository.
- [x] I HAVE NOT added vulnerabilities to the repository.
- [x] I HAVE discussed my solution with (other) members of the RA2CE team.

## What has been done?
- Modified `AnalysisResultWrapper` so that it does not reference back to `AnalysisProtocol`.
- Introduced `AnalysisResultWrapperProtocol` and new specialization `DamagesResultWrapper`.
- Adapted all concrete classes of `AnalysisProtocol` to return an `AnalysisResultWrapper`.
- Introduced new base class (`AnalysisBase`) which inherits from `AnalysisProtocol` and added to all the rest of concrete implementations so we can define generic methods there (such as converting the result).

### Checklist
- [x] Code is formatted using our custom `black` and `isort` definitions.
- [x] Tests are either added or updated.
- [x] Branch is up to date with `master`.
- [x] Updated documentation if needed.

## Additional Notes (optional)
...

### CODE REVIEW

File: tests/analysis/test_analysis_result_wrapper.py [hunk 1]
```diff
@@ -1,36 +1,96 @@
...
 
-from ra2ce.analysis.analysis_result_wrapper import AnalysisResultWrapper
+from ra2ce.analysis.analysis_result.analysis_result_wrapper import (
+    AnalysisResult,
+    AnalysisResultWrapper,
+)
 
 
-class TestAnalysisResultWrapper:
+class TestAnalysisResult:
```
Comment (ArdtK) [inline tests/analysis/test_analysis_result_wrapper.py:12]: Should be in a separate module.

**Review by ArdtK — CHANGES_REQUESTED** [review]

File: ra2ce/analysis/analysis_result/analysis_result.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+
+@dataclass(kw_only=True)
+class AnalysisResult:
+    """
+    Dataclass to represent an analysis result (`GeoDataFrame`) and its related configuration.
+    """
+
+    analysis_result: GeoDataFrame
+    analysis_config: AnalysisSectionLosses | AnalysisSectionDamages | AnalysisSectionAdaptation
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply ra2ce/analysis/analysis_result/analysis_result.py:41]: It was not available at the time of coding 😉

File: tests/analysis/conftest.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            analysis_config=mocked_analysis.analysis,
+            output_path=mocked_analysis.output_path,
+        )
+    )
+    _single_result = _result_wrapper.results_collection[0]
+    assert isinstance(_single_result, AnalysisResult)
+
+    if _single_result.output_path.exists():
+        shutil.rmtree(_single_result.output_path)
+    return _result_wrapper
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply tests/analysis/conftest.py:89]: No, yielding will return an iterator (I need to change the type hint), which means that the result wrapper will not be instantiated during the test. In this case yield will not work (you can test it yourself as well).

File: ra2ce/analysis/losses/optimal_route_origin_closest_destination.py [hunk 4]
```diff
@@ -111,5 +113,5 @@ def _save_gpkg_analysis(
             )
             del opt_routes["geometry"]
             opt_routes.to_csv(csv_path, index=False)
-
-        return None
+        # TODO: This does not seem correct, why were we returning None?
+        return self.generate_result_wrapper(None)
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply ra2ce/analysis/losses/optimal_route_origin_closest_destination.py:117]: Now that I read it with this description it does seem natural to remove the related code. I will create a new issue for it.

File: ra2ce/analysis/losses/multi_link_origin_closest_destination.py [hunk 5]
```diff
@@ -168,4 +170,5 @@ def _save_gpkg_analysis(
                 index=False,
             )
 
-        return None
+        # TODO: This does not seem correct, why were we returning None?
+        return self.generate_result_wrapper(None)
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply ra2ce/analysis/losses/multi_link_origin_closest_destination.py:174]: https://github.com/Deltares/ra2ce/issues/632

File: ra2ce/analysis/losses/optimal_route_origin_closest_destination.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Carsopre) [reply ra2ce/analysis/losses/optimal_route_origin_closest_destination.py:117]: Will be solved  in https://github.com/Deltares/ra2ce/issues/632

File: ra2ce/analysis/analysis_factory.py [hunk 6]
```diff
@@ -87,7 +87,9 @@ def get_damages_analysis(
         )
 
         if analysis.analysis == AnalysisDamagesEnum.DAMAGES:
-            return Damages(_analysis_input)
+            return Damages(
+                _analysis_input, analysis_config.graph_files.base_graph_hazard.graph
+            )
```
[the remark this answers is not in the document]
Comment (Carsopre) [reply ra2ce/analysis/analysis_factory.py:92]: I do not see at the moment the need to refactor this further. In fact we should revise what is actually the purpose of the `graph_files`, and whether we can indeed streamline it for all t he analysis (providing a subset of them depending on the analysis or so).

File: ra2ce/analysis/damages/damages.py [hunk 7]
```diff
@@ -140,7 +140,7 @@ def generate_result_wrapper(
     ) -> AnalysisResultWrapper:
         """
         Overloading of the paranet `generate_result_wrapper` to convert the
-        analysis_result which is segmented base into a link base and return
+        analysis_result which is segment base into a link base and return
```
Comment (ArdtK) [inline ra2ce/analysis/damages/damages.py:143]:
```suggestion
        analysis_result which is segment based into link based and return
```

File: ra2ce/analysis/damages/damages.py [hunk 8]
```diff
@@ -118,10 +127,177 @@ def _rename_road_gdf_to_conventions(road_gdf_columns: list[str]) -> list[str]:
...
             "The hazard calculation does not know what to do if the analysis specifies {}".format(
                 self.analysis.event_type
             )
         )
+
+    def generate_result_wrapper(
+        self, *analyses_results: GeoDataFrame
+    ) -> AnalysisResultWrapper:
+        """
+        Overloading of the paranet `generate_result_wrapper` to convert the
```
Comment (ArdtK) [inline ra2ce/analysis/damages/damages.py:142]:
```suggestion
        Overloading of the parent `generate_result_wrapper` to convert the
```

**Review by ArdtK — APPROVED** [review]
2 minor comments

**Pull request merged**